File: apps/android/app/src/main/java/com/nanolink/app/ui/design/NanoCharts.kt
package com.nanolink.app.ui.design

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.TextMeasurer
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/// Tiny sparkline used inside KPI tiles. Ports `NanoSparkline`.
@Composable
fun NanoSparkline(
    data: List<Double>,
    color: Color,
    width: Dp = 56.dp,
    height: Dp = 18.dp,
    fill: Boolean = false,
    modifier: Modifier = Modifier,
) {
    if (data.isEmpty()) {
        Spacer(modifier.size(width, height))
        return
    }
    Canvas(modifier = modifier.size(width, height)) {
        val low = data.min()
        val high = data.max()
        val range = if (high - low == 0.0) 1.0 else high - low
        val step = size.width / max(data.size - 1, 1)
        val path = Path()
        data.forEachIndexed { index, value ->
            val x = index * step
            val y = size.height - ((value - low) / range).toFloat() * (size.height - 2f) - 1f
            if (index == 0) path.moveTo(x, y) else path.lineTo(x, y)
        }
        if (fill) {
            val area = Path().apply {
                addPath(path)
                lineTo(size.width, size.height)
                lineTo(0f, size.height)
                close()
            }
            drawPath(area, color.copy(alpha = 0.12f))
        }
        drawPath(
            path = path,
            color = color,
            style = Stroke(width = 1.5f * density, cap = StrokeCap.Round, join = StrokeJoin.Round),
        )
    }
}

/// One data series for `NanoLineChart`.
data class NanoSeries(
    val data: List<Double>,
    val color: Color,
    val fill: Boolean = false,
    val dashed: Boolean = false,
    val label: String? = null,
    /// Optional per-sample peak values; shaded as a faint envelope band.
    val band: List<Double>? = null,
)

data class NanoThreshold(
    val value: Double,
    val color: Color,
    val label: String? = null,
)

/// Touch-tuned line chart used by the history graphs. Ports `NanoLineChart`.
@Composable
fun NanoLineChart(
    series: List<NanoSeries>,
    modifier: Modifier = Modifier,
    height: Dp = 140.dp,
    yMax: Double? = null,
    yMin: Double = 0.0,
    unit: String = "%",
    thresholds: List<NanoThreshold> = emptyList(),
    grid: Boolean = true,
    xLabels: List<String> = listOf("-60m", "-30m", "now"),
    timesMillis: List<Long> = emptyList(),
) {
    val t = nano
    val measurer = rememberTextMeasurer()
    val labels = if (timesMillis.size >= 2) labelsFromTimes(timesMillis) else xLabels
    val gridColor = t.sep2
    val axisColor = t.fg4

    Canvas(modifier = modifier.fillMaxWidth().height(height)) {
        val padL = 30f * density
        val padR = 6f * density
        val padT = 6f * density
        val padB = 18f * density
        val dataMax = series.flatMap { it.data }.maxOrNull() ?: 1.0
        val top = yMax ?: max(dataMax * 1.1, 1.0)
        val length = series.maxOfOrNull { it.data.size } ?: 0
        val innerW = size.width - padL - padR
        val innerH = size.height - padT - padB

        fun xFor(index: Int): Float = padL + index.toFloat() / max(length - 1, 1) * innerW
        fun yFor(value: Double): Float {
            val denominator = if (top - yMin == 0.0) 1.0 else top - yMin
            return padT + (1f - ((value - yMin) / denominator).toFloat()) * innerH
        }

        val dash = PathEffect.dashPathEffect(floatArrayOf(2f * density, 3f * density))

        if (grid) {
            for (i in 0..3) {
                val tick = yMin + (top - yMin) * i / 3
                val y = yFor(tick)
                drawLine(
                    color = gridColor,
                    start = Offset(padL, y),
                    end = Offset(size.width - padR, y),
                    strokeWidth = density,
                    pathEffect = if (i == 0 || i == 3) null else dash,
                )
                drawAxisText(
                    measurer = measurer,
                    text = "${tick.roundToInt()}$unit",
                    anchor = Offset(padL - 4f * density, y),
                    color = axisColor,
                    align = TextAlign.End,
                    density = density,
                )
            }
        }

        thresholds.forEach { threshold ->
            val y = yFor(threshold.value)
            drawLine(
                color = threshold.color.copy(alpha = 0.7f),
                start = Offset(padL, y),
                end = Offset(size.width - padR, y),
                strokeWidth = density,
                pathEffect = dash,
            )
            threshold.label?.let { label ->
                drawAxisText(
                    measurer = measurer,
                    text = label,
                    anchor = Offset(size.width - padR - 2f * density, y - 6f * density),
                    color = threshold.color,
                    align = TextAlign.End,
                    density = density,
                )
            }
        }

        // max-envelope bands
        series.forEach { entry ->
            val band = entry.band ?: return@forEach
            if (band.isEmpty() || entry.data.isEmpty()) return@forEach
            val count = min(band.size, entry.data.size)
            if (count < 2) return@forEach
            val area = Path()
            for (i in 0 until count) {
                val x = xFor(i)
                val y = yFor(max(band[i], entry.data[i]))
                if (i == 0) area.moveTo(x, y) else area.lineTo(x, y)
            }
            for (i in count - 1 downTo 0) {
                area.lineTo(xFor(i), yFor(entry.data[i]))
            }
            area.close()
            drawPath(area, entry.color.copy(alpha = 0.10f))
        }

        series.forEach { entry ->
            if (entry.data.isEmpty()) return@forEach
            val path = Path()
            entry.data.forEachIndexed { index, value ->
                val x = xFor(index)
                val y = yFor(value)
                if (index == 0) path.moveTo(x, y) else path.lineTo(x, y)
            }
            if (entry.fill) {
                val area = Path().apply {
                    addPath(path)
                    lineTo(xFor(entry.data.size - 1), yFor(yMin))
                    lineTo(xFor(0), yFor(yMin))
                    close()
                }
                drawPath(area, entry.color.copy(alpha = 0.15f))
            }
            drawPath(
                path = path,
                color = entry.color,
                style = Stroke(
                    width = 1.5f * density,
                    cap = StrokeCap.Round,
                    join = StrokeJoin.Round,
                    pathEffect = if (entry.dashed) {
                        PathEffect.dashPathEffect(floatArrayOf(4f * density, 3f * density))
                    } else {
                        null
                    },
                ),
            )
        }

        val ticks = if (labels.size >= 3) labels else listOf("", "", "now")
        for (i in 0..2) {
            val x = padL + i / 2f * innerW
            val align = when (i) {
                0 -> TextAlign.Start
                2 -> TextAlign.End
                else -> TextAlign.Center
            }
            drawAxisText(
                measurer = measurer,
                text = ticks[i],
                anchor = Offset(x, size.height - 8f * density),
                color = axisColor,
                align = align,
                density = density,
            )
        }
    }
}

private enum class TextAlign { Start, Center, End }

private fun DrawScope.drawAxisText(
    measurer: TextMeasurer,
    text: String,
    anchor: Offset,
    color: Color,
    align: TextAlign,
    density: Float,
) {
    if (text.isEmpty()) return
    val layout = measurer.measure(
        text = text,
        style = TextStyle(fontSize = 9.sp, fontFamily = NanoMonoFamily, color = color),
    )
    val width = layout.size.width.toFloat()
    val heightPx = layout.size.height.toFloat()
    val x = when (align) {
        TextAlign.Start -> anchor.x
        TextAlign.Center -> anchor.x - width / 2
        TextAlign.End -> anchor.x - width
    }
    drawText(layout, topLeft = Offset(x, anchor.y - heightPx / 2))
}

/// Derive three tick labels (start / middle / now) from real timestamps.
fun labelsFromTimes(timesMillis: List<Long>): List<String> {
    if (timesMillis.size < 2) return listOf("", "", "now")
    val last = timesMillis.last()
    fun relative(value: Long): String {
        val seconds = (last - value) / 1_000
        return when {
            seconds <= 30 -> "now"
            seconds < 3_600 -> "-${(seconds / 60.0).roundToInt()}m"
            seconds < 86_400 -> "-${(seconds / 3_600.0).roundToInt()}h"
            else -> "-${(seconds / 86_400.0).roundToInt()}d"
        }
    }
    return listOf(
        relative(timesMillis.first()),
        relative(timesMillis[timesMillis.size / 2]),
        relative(last),
    )
}

/// Circular gauge (CPU/MEM donuts). Ports `NanoDonut`.
@Composable
fun NanoDonut(
    value: Double,
    label: String,
    maxValue: Double = 100.0,
    size: Dp = 64.dp,
    thickness: Dp = 5.dp,
    sub: String? = null,
    tone: Color? = null,
    modifier: Modifier = Modifier,
) {
    val t = nano
    val fraction = (value / maxValue).coerceIn(0.0, 1.0)
    val color = tone ?: when {
        fraction > 0.9 -> t.crit
        fraction > 0.75 -> t.warn
        else -> t.fg
    }
    val trackColor = t.card3
    Box(modifier = modifier.size(size), contentAlignment = Alignment.Center) {
        Canvas(Modifier.size(size)) {
            val strokePx = thickness.toPx()
            val inset = strokePx * 1.5f
            val arcSize = Size(this.size.width - inset * 2, this.size.height - inset * 2)
            drawArc(
                color = trackColor,
                startAngle = 0f,
                sweepAngle = 360f,
                useCenter = false,
                topLeft = Offset(inset, inset),
                size = arcSize,
                style = Stroke(width = strokePx),
            )
            drawArc(
                color = color,
                startAngle = -90f,
                sweepAngle = (360 * fraction).toFloat(),
                useCenter = false,
                topLeft = Offset(inset, inset),
                size = arcSize,
                style = Stroke(width = strokePx, cap = StrokeCap.Round),
            )
        }
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text(
                text = label,
                fontSize = if (size > 70.dp) 16.sp else 13.sp,
                fontWeight = FontWeight.SemiBold,
                color = t.fg,
            )
            if (sub != null) {
                Text(sub, fontSize = 9.sp, fontFamily = NanoMonoFamily, color = t.fg4)
            }
        }
    }
}

/// Per-core utilization matrix (vertical bars in a grid). Ports `NanoCoreMatrix`.
@Composable
fun NanoCoreMatrix(
    cores: List<Double>,
    columns: Int = 8,
    modifier: Modifier = Modifier,
) {
    val t = nano
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(3.dp)) {
        cores.chunked(columns).forEach { row ->
            Row(horizontalArrangement = Arrangement.spacedBy(3.dp), modifier = Modifier.fillMaxWidth()) {
                row.forEach { value ->
                    val tone = when {
                        value > 90 -> t.crit
                        value > 70 -> t.warn
                        value > 30 -> t.fg2
                        else -> t.fg4
                    }
                    Box(
                        modifier = Modifier
                            .weight(1f)
                            .aspectRatio(1.6f)
                            .clip(RoundedCornerShape(2.dp))
                            .background(t.card3),
                        contentAlignment = Alignment.BottomCenter,
                    ) {
                        Box(
                            modifier = Modifier
                                .fillMaxWidth()
                                .fillMaxHeight((value / 100).coerceIn(0.0, 1.0).toFloat())
                                .background(tone.copy(alpha = 0.85f)),
                        )
                    }
                }
                // Keep the last partial row aligned with the full rows above it.
                repeat(columns - row.size) {
                    Spacer(Modifier.weight(1f))
                }
            }
        }
    }
}

/// KPI summary tile used on the dashboard. Ports `NanoKpiTile`.
@Composable
fun NanoKpiTile(
    label: String,
    value: String,
    sub: String,
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    modifier: Modifier = Modifier,
    tone: String? = null,
    spark: List<Double>? = null,
    background: Color? = null,
) {
    val t = nano
    val toneColor = when (tone) {
        "crit" -> t.crit
        "warn" -> t.warn
        else -> t.accent
    }
    val valueColor = when (tone) {
        "crit" -> t.crit
        "warn" -> t.warn
        else -> t.fg
    }
    Column(
        modifier = modifier
            .clip(RoundedCornerShape(t.cardRadius))
            .background(background ?: t.card)
            .padding(horizontal = 14.dp, vertical = 12.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(5.dp)) {
            androidx.compose.material3.Icon(
                icon,
                contentDescription = null,
                tint = t.fg3,
                modifier = Modifier.size(14.dp),
            )
            Text(label, fontSize = 12.sp, fontWeight = FontWeight.Medium, color = t.fg3)
        }
        Spacer(Modifier.height(4.dp))
        Text(
            text = value,
            fontSize = 26.sp,
            fontWeight = FontWeight.SemiBold,
            letterSpacing = (-0.3).sp,
            color = valueColor,
        )
        Spacer(Modifier.height(2.dp))
        Row(verticalAlignment = Alignment.Bottom) {
            Text(
                text = sub,
                fontSize = 10.5.sp,
                fontFamily = NanoMonoFamily,
                color = t.fg4,
                maxLines = 1,
                modifier = Modifier.weight(1f, fill = false),
            )
            Spacer(Modifier.weight(1f))
            if (spark != null) {
                NanoSparkline(data = spark, color = toneColor)
            }
        }
    }
}

/// Row: leading icon + short label + meter + right-aligned value/sub.
/// Ports `NanoMetricRow`.
@Composable
fun NanoMetricRow(
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    label: String,
    value: String,
    modifier: Modifier = Modifier,
    sub: String? = null,
    percent: Double? = null,
    tone: String? = null,
) {
    val t = nano
    val toneColor = when (tone) {
        "crit" -> t.crit
        "warn" -> t.warn
        else -> t.fg
    }
    Row(modifier = modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        androidx.compose.material3.Icon(
            icon,
            contentDescription = null,
            tint = t.fg4,
            modifier = Modifier.size(13.dp),
        )
        Spacer(Modifier.width(10.dp))
        Text(
            text = label,
            fontSize = 11.sp,
            fontFamily = NanoMonoFamily,
            color = t.fg3,
            modifier = Modifier.width(30.dp),
        )
        Spacer(Modifier.width(10.dp))
        Box(Modifier.weight(1f)) {
            if (percent != null) {
                NanoMeter(value = percent / 100, color = if (tone == null) null else toneColor)
            }
        }
        Spacer(Modifier.width(10.dp))
        Column(horizontalAlignment = Alignment.End, modifier = Modifier.width(64.dp)) {
            Text(value, fontSize = 13.sp, fontWeight = FontWeight.Medium, color = toneColor)
            if (sub != null) {
                Text(sub, fontSize = 10.sp, fontFamily = NanoMonoFamily, color = t.fg4)
            }
        }
    }
}

File: apps/ios/NanoLink/Design/NanoCharts.swift
import SwiftUI

// MARK: - Sparkline

/// Tiny sparkline used inside KPI tiles. Ports `NanoSparkline`.
struct NanoSparkline: View {
    let data: [Double]
    let color: Color
    var width: CGFloat = 56
    var height: CGFloat = 18
    var fill: Bool = false

    var body: some View {
        Canvas { ctx, size in
            guard !data.isEmpty else { return }
            let lo = data.min() ?? 0
            let hi = data.max() ?? 1
            let range = (hi - lo) == 0 ? 1 : (hi - lo)
            let step = size.width / CGFloat(max(data.count - 1, 1))
            var path = Path()
            for i in 0..<data.count {
                let x = CGFloat(i) * step
                let y = size.height - CGFloat((data[i] - lo) / range) * (size.height - 2) - 1
                if i == 0 { path.move(to: CGPoint(x: x, y: y)) }
                else { path.addLine(to: CGPoint(x: x, y: y)) }
            }
            if fill {
                var area = path
                area.addLine(to: CGPoint(x: size.width, y: size.height))
                area.addLine(to: CGPoint(x: 0, y: size.height))
                area.closeSubpath()
                ctx.fill(area, with: .color(color.opacity(0.12)))
            }
            ctx.stroke(path, with: .color(color),
                       style: StrokeStyle(lineWidth: 1.5, lineCap: .round, lineJoin: .round))
        }
        .frame(width: width, height: height)
    }
}

// MARK: - Line chart

/// One data series for `NanoLineChart`.
struct NanoSeries {
    let data: [Double]
    let color: Color
    var fill: Bool = false
    var dashed: Bool = false
    var label: String? = nil
    /// Optional per-sample peak values; shaded as a faint envelope band.
    var band: [Double]? = nil
}

struct NanoThreshold {
    let value: Double
    let color: Color
    var label: String? = nil
}

/// Touch-tuned line chart (history graphs). Ports `NanoLineChart` / `MLineChart`.
struct NanoLineChart: View {
    let series: [NanoSeries]
    var height: CGFloat = 140
    var yMax: Double? = nil
    var yMin: Double = 0
    var unit: String = "%"
    var thresholds: [NanoThreshold] = []
    var grid: Bool = true
    var xLabels: [String] = ["-60m", "-30m", "now"]
    var times: [Date] = []
    @Environment(\.nano) private var t

    private static let padL: CGFloat = 30, padR: CGFloat = 6, padT: CGFloat = 6, padB: CGFloat = 18

    /// Derive three tick labels (start / middle / now) from real timestamps.
    static func labelsFromTimes(_ times: [Date]) -> [String] {
        guard times.count >= 2 else { return ["", "", "now"] }
        let first = times.first!
        let last = times.last!
        let mid = times[times.count / 2]
        func rel(_ d: Date) -> String {
            let s = Int(last.timeIntervalSince(d))
            if s <= 30 { return "now" }
            if s < 3600 { return "-\(Int((Double(s) / 60).rounded()))m" }
            if s < 86400 { return "-\(Int((Double(s) / 3600).rounded()))h" }
            return "-\(Int((Double(s) / 86400).rounded()))d"
        }
        return [rel(first), rel(mid), rel(last)]
    }

    var body: some View {
        let labels = times.count >= 2 ? Self.labelsFromTimes(times) : xLabels
        let gridColor = t.sep2
        let axisText = t.fg4
        Canvas { ctx, size in
            let padL = Self.padL, padR = Self.padR, padT = Self.padT, padB = Self.padB
            let allVals = series.flatMap { $0.data }
            let dataMax = allVals.max() ?? 1
            let ymax = yMax ?? max(dataMax * 1.1, 1)
            let ymin = yMin
            let len = series.reduce(0) { max($0, $1.data.count) }
            let innerW = size.width - padL - padR
            let innerH = size.height - padT - padB
            func xFor(_ i: Int) -> CGFloat { padL + CGFloat(i) / CGFloat(max(len - 1, 1)) * innerW }
            func yFor(_ v: Double) -> CGFloat {
                let denom = (ymax - ymin) == 0 ? 1 : (ymax - ymin)
                return padT + (1 - CGFloat((v - ymin) / denom)) * innerH
            }

            func text(_ s: String, at p: CGPoint, color: Color, align: HorizontalAlignment = .leading) {
                let resolved = ctx.resolve(Text(s).font(NanoFont.mono(9)).foregroundColor(color))
                let sz = resolved.measure(in: CGSize(width: 200, height: 20))
                var dx = p.x
                if align == .trailing { dx -= sz.width }
                if align == .center { dx -= sz.width / 2 }
                ctx.draw(resolved, at: CGPoint(x: dx + sz.width / 2, y: p.y))
            }

            // grid + y labels
            if grid {
                for i in 0...3 {
                    let tv = ymin + (ymax - ymin) * Double(i) / 3
                    let y = yFor(tv)
                    var line = Path()
                    line.move(to: CGPoint(x: padL, y: y))
                    line.addLine(to: CGPoint(x: size.width - padR, y: y))
                    if i == 0 || i == 3 {
                        ctx.stroke(line, with: .color(gridColor), lineWidth: 1)
                    } else {
                        ctx.stroke(line, with: .color(gridColor),
                                   style: StrokeStyle(lineWidth: 1, dash: [2, 3]))
                    }
                    text("\(Int(tv.rounded()))\(unit)", at: CGPoint(x: padL - 4, y: y),
                         color: axisText, align: .trailing)
                }
            }

            // thresholds
            for th in thresholds {
                let y = yFor(th.value)
                var line = Path()
                line.move(to: CGPoint(x: padL, y: y))
                line.addLine(to: CGPoint(x: size.width - padR, y: y))
                ctx.stroke(line, with: .color(th.color.opacity(0.7)),
                           style: StrokeStyle(lineWidth: 1, dash: [2, 3]))
                if let label = th.label {
                    text(label, at: CGPoint(x: size.width - padR - 2, y: y - 6),
                         color: th.color, align: .trailing)
                }
            }

            // max-envelope bands
            for s in series {
                guard let band = s.band, !band.isEmpty, !s.data.isEmpty else { continue }
                let n = min(band.count, s.data.count)
                if n < 2 { continue }
                var area = Path()
                for i in 0..<n {
                    let x = xFor(i)
                    let y = yFor(max(band[i], s.data[i]))
                    if i == 0 { area.move(to: CGPoint(x: x, y: y)) }
                    else { area.addLine(to: CGPoint(x: x, y: y)) }
                }
                for i in stride(from: n - 1, through: 0, by: -1) {
                    area.addLine(to: CGPoint(x: xFor(i), y: yFor(s.data[i])))
                }
                area.closeSubpath()
                ctx.fill(area, with: .color(s.color.opacity(0.10)))
            }

            // series
            for s in series {
                guard !s.data.isEmpty else { continue }
                var path = Path()
                for i in 0..<s.data.count {
                    let x = xFor(i), y = yFor(s.data[i])
                    if i == 0 { path.move(to: CGPoint(x: x, y: y)) }
                    else { path.addLine(to: CGPoint(x: x, y: y)) }
                }
                if s.fill {
                    var area = path
                    area.addLine(to: CGPoint(x: xFor(s.data.count - 1), y: yFor(ymin)))
                    area.addLine(to: CGPoint(x: xFor(0), y: yFor(ymin)))
                    area.closeSubpath()
                    ctx.fill(area, with: .color(s.color.opacity(0.15)))
                }
                let style = StrokeStyle(lineWidth: 1.5, lineCap: .round, lineJoin: .round,
                                        dash: s.dashed ? [4, 3] : [])
                ctx.stroke(path, with: .color(s.color), style: style)
            }

            // x labels
            let xl = labels.count >= 3 ? labels : ["", "", "now"]
            for i in 0..<3 {
                let x = padL + CGFloat(i) / 2 * innerW
                let align: HorizontalAlignment = i == 0 ? .leading : (i == 2 ? .trailing : .center)
                text(xl[i], at: CGPoint(x: x, y: size.height - 8), color: axisText, align: align)
            }
        }
        .frame(height: height)
        .frame(maxWidth: .infinity)
    }
}

// MARK: - Donut

/// Circular gauge (CPU/MEM donuts). Ports `NanoDonut`.
struct NanoDonut: View {
    let value: Double
    var max: Double = 100
    var size: CGFloat = 64
    var thickness: CGFloat = 5
    let label: String
    var sub: String? = nil
    var tone: Color? = nil
    @Environment(\.nano) private var t

    var body: some View {
        let pct = min(Swift.max(value / max, 0), 1)
        let color = tone ?? (pct > 0.9 ? t.crit : (pct > 0.75 ? t.warn : t.fg))
        ZStack {
            Circle()
                .stroke(t.card3, lineWidth: thickness)
                .padding(thickness)
            Circle()
                .trim(from: 0, to: CGFloat(pct))
                .stroke(color, style: StrokeStyle(lineWidth: thickness, lineCap: .round))
                .rotationEffect(.degrees(-90))
                .padding(thickness)
            VStack(spacing: 0) {
                Text(label)
                    .font(.system(size: size > 70 ? 16 : 13, weight: .semibold))
                    .foregroundColor(t.fg)
                    .monospacedDigit()
                if let sub {
                    Text(sub).font(NanoFont.mono(9)).foregroundColor(t.fg4)
                }
            }
        }
        .frame(width: size, height: size)
    }
}

// MARK: - Core matrix

/// Per-core utilization matrix (vertical bars in a grid). Ports `NanoCoreMatrix`.
struct NanoCoreMatrix: View {
    let cores: [Double]
    var cols: Int = 8
    @Environment(\.nano) private var t

    var body: some View {
        let columns = Array(repeating: GridItem(.flexible(), spacing: 3), count: cols)
        LazyVGrid(columns: columns, spacing: 3) {
            ForEach(Array(cores.enumerated()), id: \.offset) { _, v in
                let tone: Color = v > 90 ? t.crit : (v > 70 ? t.warn : (v > 30 ? t.fg2 : t.fg4))
                GeometryReader { geo in
                    ZStack(alignment: .bottom) {
                        Rectangle().fill(t.card3)
                        Rectangle()
                            .fill(tone.opacity(0.85))
                            .frame(height: geo.size.height * CGFloat(min(Swift.max(v / 100, 0), 1)))
                    }
                }
                .aspectRatio(1.6, contentMode: .fit)
                .clipShape(RoundedRectangle(cornerRadius: 2, style: .continuous))
            }
        }
    }
}
